File: Python/200. Number of Islands.py
class Solution:
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if len(grid) == 0:
            return 0
        dx = [0, 1, -1, 0]
        dy = [1, 0, 0, -1]
        ans = 0
        n, m = len(grid), len(grid[0])
        for i in range(n):
            for j in range(m):
                if grid[i][j] == "1":
                    q = [[i,j]]
                    ans += 1
                    while len(q) > 0:
                        now = q.pop(0)[:]
                        for k in range(4):
                            tx = now[0] + dx[k]
                            ty = now[1] + dy[k]
                            if tx < 0 or ty < 0:
                                continue
                            if tx >=n or ty >= m:
                                continue
                            if grid[tx][ty] == "1":
                                grid[tx][ty] = "0"
                                q.append([tx, ty])
        return ans
                                

File: C++/200. Number of Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int dx[] = {0, 0, 1, -1};
        int dy[] = {1, -1, 0, 0};
        if(grid.size()==0)
            return 0;
        int n = grid.size(), m = grid[0].size();
        int ans = 0;
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j] == '0')
                    continue;
                ans += 1;
                queue<pair<int, int> >q;
                q.push(make_pair(i, j));
                while(!q.empty())
                {
                    pair<int, int> now = q.front();
                    q.pop();
                    for(int k=0;k<4;k++)
                    {
                        int tx = now.first + dx[k];
                        int ty = now.second + dy[k];
                        if(tx<0 || ty < 0 || tx>=n || ty>=m)
                            continue;
                        if(grid[tx][ty]=='1')
                        {
                            grid[tx][ty] = '0';
                            q.push(make_pair(tx, ty));
                        }
                    }
                }
            }
        }
        return ans;
    }
};